pkgs/development/interpreters/python/hooks/fix-python-deps.sh

```diff
@@ -0,0 +1,76 @@
+# shellcheck shell=bash
+
+# Setup hook that fixes Python dependencies versions.
+#
+# Example usage in a derivation:
+#
+#   { …, pythonPackages, … }:
+#
+#   pythonPackages.buildPythonPackage {
+#     …
+#     nativeBuildInputs = [ pythonPackages.fixPythonDepsHook ];
+#
+#     # Already the default, only change if necessary
+#     requirementsTxt = "requirements.txt";
+#
+#     # This will remove the upperbound version of the package
+#     # e.g.: abc>1<=2 -> abc>1
+#     relaxPythonDeps = [ "abc" ];
+#     # This will remove all dependencies restrictions
+#     # e.g.: cde>1<=2 -> cde
+#     unlockPythonDeps = [ "cde" ];
+#     # This will remove the dependency
+#     # e.g.: fgh>1<=2 -> <nothing>
+#     removePythonDeps = [ "cde" ];
+#     …
+#   }
+#
+# This hook will copy files which are either given by full path
+# or all '*.desktop' files placed inside the 'share/applications'
+# folder of each `desktopItems` argument.
+
+declare -a relaxPythonDeps
+declare -a unlockPythonDeps
+declare -a removePythonDeps
+
+requirementsTxt="${requirementsTxt:-requirements.txt}"
+
+relaxPythonDepsHook() {
+    echo "Executing relaxPythonDepsHook"
+
+    if [[ -z "$relaxPythonDeps" ]]; then return; fi
+    for dep in ${relaxPythonDeps[@]}; do
+        sed -r "s|${dep}([>=0-9.,]*)[ <=0-9.]*|${dep}\1|" -i "$requirementsTxt"
+    done
+}
+
+unlockPythonDepsHook() {
+    echo "Executing unlockPythonDepsHook"
+
+    if [[ -z "$unlockPythonDeps" ]]; then return; fi
+    for dep in ${unlockPythonDeps[@]}; do
+        sed -r "s|(${dep})[ ><=0-9.,]*|\1|" -i "$requirementsTxt"
+    done
+}
+
+removePythonDepsHook() {
+    echo "Executing removePythonDepsHook"
+
+    if [[ -z "$removePythonDeps" ]]; then return; fi
+    for dep in ${removePythonDeps[@]}; do
+        sed -r "s|${dep}[ ><=0-9.,]*||" -i "$requirementsTxt"
```

Comment:
That wouldn't work with something like ``xattr>=0.4;sys_platform!='win32'``?

Reply:
Yeah, it wouldn't. Also with something like `abc>=1.0b`.

Maybe I really need to write this with a proper parser, otherwise the chance of creating difficult to fix bugs is very big. I saw this one here for `requirements.txt`: https://github.com/madpah/requirements-parser (it is already in nixpkgs).

My only fear is that having a Python build step depending on Python itself can lead to some circular problems during breakage, where the failure of `requirements-parser` for example could break all other packages and make it difficult to fix. But looking at another way at least the package doesn't seem to have much dependencies: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/requirements-parser/default.nix

Not sure yet.

Reply:
Switched to a proper parse, so I think this kinda of issue shouldn't be a problem anymore.